### PR TITLE
wire should be bytes or string

### PR DESCRIPTION
**Pull request by beltran**
Fixes https://github.com/eventlet/eventlet/issues/499

### CODE REVIEW

File: tests/greendns_test.py [hunk 1]
```diff
@@ -854,3 +854,15 @@ def test_hosts_priority():
 def test_import_rdtypes_then_eventlet():
     # https://github.com/eventlet/eventlet/issues/479
     tests.run_isolated('greendns_import_rdtypes_then_eventlet.py')
+
+
+def test_nonameserver_is_risen():
```
Comment (temoto) [inline tests/greendns_test.py]: These apply to all code:

- test is named `nonameserver_is_risen` but it doesn't check for any exception
- imports inside function is a recipe for disaster, use it only as last resort
- `reset_default_resolver()` modifies global state, so it renders all other tests results unreliable

This is eventlet project specific, some don't mind: don't rely on external services. Maybe you can't access 8.8.8.8, or this address is not what you expect or it doesn't know about `github.com`.

Solutions:
- for naming - name should be consistent with code
- for imports and global state - isolated tests, see code above
- for external dependencies - mock network IO or run small DNS server that's programmed to respond what you want

**Comment by temoto** [comment]
Thank you for valuable contribution. You can fix tests or do nothing, I will write a test.

**Comment by beltran** [comment]
I will fix the test, thank you for the detailed review and sorry for the clumsy test

File: tests/greendns_test.py [hunk 2]
```diff
@@ -854,3 +857,28 @@ def test_hosts_priority():
 def test_import_rdtypes_then_eventlet():
     # https://github.com/eventlet/eventlet/issues/479
     tests.run_isolated('greendns_import_rdtypes_then_eventlet.py')
+
+
+def test_raise_dns_tcp():
+    # https://github.com/eventlet/eventlet/issues/499
+    try:
+        dns.resolver.query('host.example.com', 'a', tcp=True)
+        assert False, "Should have arisen NXDOMAIN or NoNameservers"
```
Comment (temoto) [inline tests/greendns_test.py]: Test does not ensure that `host.example.com` query to unknown default resolver results in NXDOMAIN.

File: tests/__init__.py [hunk 3]
```diff
@@ -406,3 +409,88 @@ def test_run_python_pythonpath_extend():
...
+                self.data = client_sock.recv(1024)
+                if bytearray(self.data) == b"DEATH_PILL":
+                    client_sock.close()
+                    break
+
+                self.init_domain()
+                client_sock.send(self.answer())
+                client_sock.close()
+        except socket.error:
+            pass
```
Comment (temoto) [inline tests/__init__.py]: Ignoring errors is always a bad idea.

File: tests/greendns_test.py [hunk 4]
```diff
@@ -854,3 +857,28 @@ def test_hosts_priority():
...
+    try:
+        tiny_server = tests.TinyDNSServerTCP("192.168.1.1")
+        tiny_server.start()
+        resolver.nameserver_ports["127.0.0.1"] = tiny_server.get_port()
+        response = resolver.query('host.example.com', 'a', tcp=True)
+
+        assert isinstance(response, Answer)
+    finally:
+        tiny_server.stop()
+        pass
```
Comment (temoto) [inline tests/greendns_test.py]: (minor) redundant `pass`

File: tests/greendns_test.py [hunk 5]
```diff
@@ -854,3 +857,28 @@ def test_hosts_priority():
...
+def test_raise_dns_tcp():
+    # https://github.com/eventlet/eventlet/issues/499
+    try:
+        dns.resolver.query('host.example.com', 'a', tcp=True)
+        assert False, "Should have arisen NXDOMAIN or NoNameservers"
+    except (dns.resolver.NXDOMAIN, dns.resolver.NoNameservers):
+        pass
+
+
+def test_noraise_dns_tcp():
```
Comment (temoto) [inline tests/greendns_test.py]: - `TinyDNSServerTCP` listens `localhost` (and there is no way to change it and that's fine)
- test goes to `127.0.0.1` which is not the same as localhost
- `127...` string duplicate
- `192.168.1.1` is not used
- response is not checked against `192...`

**Comment by temoto** [comment]
On DNS server design. I understand it works, feel free to leave it as is.

Here's a simpler version:
```
@contextlib.contextmanager
def dns_tcp_server(request_count=1):
  response = random_ip()
  state = [0]  # request count storage writable by thread

  def check(x):
    if isinstance(x, dns.resolver.Answer):
      assert response in x
    else:
      assert x == response

  def thread_fun():  # thread target
    while True:
      accept, recv, sendall
      # note socket.send is not enough https://docs.python.org/2/library/socket.html#socket.socket.send
      state[0] += 1
      if state[0] >= request_count:
        return

  sock = create, bind, listen
  server_addr = sock.getsocketname()
  thread = create, start
  yield server_addr, check
  thread.join()  # forever, all tests run with some form of timeout control

def test_dns_tcp():
    # https://github.com/eventlet/eventlet/issues/499
    with dns_tcp_server() as s:
        dnsaddr, check = s
        resolver = Resolver()
        resolver.nameservers = [dnsaddr[0]]
        resolver.nameserver_ports[dnsaddr[0]] = dnsaddr[1]
        response = resolver.query('host.example.com', 'a', tcp=True)
        check(response)
```

How is this better?
- no classes, instant win :-)
- no Event synchronisation, there was no reason to delay thread start and socket listen, so port is available at creation
- you may forget to do `server.stop()` and it still works
- check value of randomized response
- address of DNS server is not defined in test

**Comment by beltran** [comment]
Thank you very much for the review @temoto, I think I've addressed all your comments.

**Comment by temoto** [comment]
The purpose of request_count was to replace death_pill "end cleaner" with strict contract: test specifies how many requests it intends to send, receiving more or less is an error in eventlet greendns middleware or in test code.

But I'm not going to test your patience. :-) Thank you very much for your work.

**Comment by beltran** [comment]
Ohh I see, I thought it was throttling the max number of connections. I could've put up with that :muscle:

**Pull request merged**